editions/tw5.com/tiddlers/widgets/LinkWidget.tid

```diff
@@ -54,12 +55,20 @@ This causes the tooltip to be the ''tooltip'' field of the target tiddler. If th
 
 ! CSS Classes
 
+The link widget automatically determines and applies the following classes to links:
+
 * `tc-tiddlylink` - applied to all links
 * `tc-tiddlylink-external` - applied to external, non-tiddler links
 * `tc-tiddlylink-internal` - applied to tiddler links
 * `tc-tiddlylink-missing` - applied to tiddler links where the target tiddler doesn't exist
 * `tc-tiddlylink-resolves` - applied to tiddler links when the target tiddler does exist
 
+Use the `class` attribute to specify additional classes, or use the prefix `!` to only apply those classes, e.g. when used in a LinkCatcherWidget:
+
+<<wikitext-example-without-html """*<$link class="example">Here</$link> the `example` class is added.
+*<$link class="!example">Here</$link> only the `example` class applies.
+*<$link class="!">Here</$link> no class is set.""">>
```

Comment:
Hmmm, 

If we have a look to `class="!something"` from a HTML perspective, it is undefined syntax. .... Why don't we use: 

`<$link enforceClass="example">Here</$link>` ... IMO it would also simplify the code itself.

Reply:
Why would it be faster? We are not defining widgets from a html perspective. If we did so, then we could also never do `foo=<<bar>>` or foo=`{{!!bar}}`.

The additional attribute, to me is overkill and complicates stuff. I want to define a class, not an *enforceClass*... and an exclamation mark is - by intention - an otherwise invalid class. So, if you see that, it automatically makes "click" that here we must have something else going on other than a simple class.

Reply:
>The additional attribute, to me is overkill ...

Why overkill. It would be even less code, which imo will be much simpler to maintain. 

>I want to define a class, 

If you want to define a class, you should use valid class values.

Reply:
> It would be even less code, which imo will be much simpler to maintain.

Because it complicates both documentation and implementation. What if I define **enforeClass** and also **class**? I guess you would say **enforceClass** is what applies then. Is it?

> If you want to define a class, you should use valid class values.

I am, the class value is that which follows the `!` The exclamation mark simply serves as a flag or prefix, as does the `+` in `+[limit[10]]`. It's simply a question of convention. I don't see anything wrong about either convention.

Yours makes it explicit, but yields confusion as to when **class** and when **enforceClass** apply.

Mine may look a bit awkward. But it does so by design, with intention.

As for peformance, I only believe it when you prove it, that doing a substr is more costly than parsing an entire attribute.

Reply:
I think, using a new attribute would make the whole code simpler

Reply:
@pmario: I don't think anything has become simpler. However, I also don't really mind. So, I updated the PR.